src/SortingAlgorithms.jl

```diff
@@ -652,4 +681,239 @@ else
+    while f(page)
+        plc = pageLocations[page-3] # plc has data belonging to page
+        pageLocations[page-3] = page
+        copyto!(v, page_offset(page) + 1, v, page_offset(plc) + 1, pagesize)
+        page = plc
+    end
+    page
+end
+
+# merge v[lo:m] (A) and v[m+1:hi] (B) using scratch[] in O(sqrt(n)) space
+function paged_merge!(v::AbstractVector{T}, lo::Integer, m::Integer, hi::Integer, o::Ordering, scratch::AbstractVector{T}, pageLocations::AbstractVector{<:Integer}) where {T}
+    @assert lo < m < hi
+    lenA = 1 + m - lo
+    lenB = hi - m
+
+    # this function only supports merges with length(A) <= length(B),
+    # which is guaranteed by pagedmergesort!
+    @assert lenA <= lenB
+
+    # regular merge if scratch is big enough
+    lenA <= length(scratch) && return merge!(v, lo, m, hi, o, scratch)
+
+    len = lenA + lenB
+    pagesize = isqrt(len)
+    nPages = len ÷ pagesize # a partial page at the end does not count
+    @assert length(scratch) >= 3pagesize
+    @assert length(pageLocations) >= nPages - 3
+
+    @inline page_offset(page) = (page - 1) * pagesize + lo - 1
+
+    @inbounds begin
+        ##################
+        # merge
+        ##################
+        # merge the first 3 pages into scratch
+        a, b, _ = merge!((_, _, k) -> k <= 3pagesize, scratch, v, v, o, lo, m + 1, 1)
+        # initialize variables for merging into pages
+        pages = Pages(-17, 0, 1, (m - lo) ÷ pagesize + 2) # first argument is unused
+        # more efficient loop while more than pagesize elements of A and B are remaining
+        while_condition1(offset) = (_, _, k) -> k <= offset + pagesize
+        while a < m - pagesize && b < hi - pagesize
+            pages = next_page!(pageLocations, pages, pagesize, lo, a)
+            offset = page_offset(pages.current)
+            a, b, _ = merge!(while_condition1(offset), v, v, v, o, a, b, offset + 1)
+        end
+        # merge until either A or B is empty or the last page is reached
+        k, offset = nothing, nothing
+        while_condition2(offset) = (a, b, k) -> k <= offset + pagesize && a <= m && b <= hi
+        while a <= m && b <= hi && pages.currentNumber + 3 < nPages
+            pages = next_page!(pageLocations, pages, pagesize, lo, a)
+            offset = page_offset(pages.current)
+            a, b, k = merge!(while_condition2(offset), v, v, v, o, a, b, offset + 1)
+        end
+        # if the last page is reached, merge the remaining elements into the final partial page
+        if pages.currentNumber + 3 == nPages && a <= m && b <= hi
+            a, b, k = merge!((a, b, _) -> a <= m && b <= hi, v, v, v, o, a, b, nPages * pagesize + lo)
+            copyto!(v, k, v, a <= m ? a : b, hi - k + 1)
+        else
+            use_a = a <= m
+            # copy the incomplete page
+            partial_page_size = offset + pagesize - k + 1
+            copyto!(v, k, v, use_a ? a : b, partial_page_size)
```

Comment:
It turns out I was wrong when I said it was bad to reinvent the wheel with `copyto!`. `Base.copyto!` is [buggy](https://github.com/JuliaLang/julia/issues/50900). Because it tries to detect aliasing which is not really possible to do effectively in Julia.

We should roll our own :`(
```julia
function _unsafe_copyto!(dest, doffs, src, soffs, n)
    @inbounds for i in 0:n-1
        dest[doffs + i] = src[soffs + i]
    end
    dest
end

function _unsafe_copyto!(dest::Array, doffs, src::Array, soffs, n)
    unsafe_copyto!(dest, doffs, src, soffs, n)
end
```

Reply:
> It turns out I was wrong when I said it was bad to reinvent the wheel [...].

Good catch! But beeing right twice does not mean you were wrong. :wink: Using copyto! definitively led to better code. Even after using the custom unsafe version.